VideoWeb/VideoWeb.ChatHub/Hub/ChatHub.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Logging;
+using VideoWeb.Common.SignalR;
+using VideoWeb.Services.Video;
+
+namespace VideoWeb.ChatHub.Hub
+{
+    public interface IChatHubClient
+    {
+        Task ReceiveMessage(Guid conferenceId, string from, string message, DateTime timestamp);
+    }
+
+    [Authorize(Policy = "EventHubUser")]
+    public class ChatHub : Hub<IChatHubClient>
+    {
+        private readonly IUserProfileService _userProfileService;
+        private readonly IVideoApiClient _videoApiClient;
+
+        private readonly ILogger<ChatHub> _logger;
+
+        public ChatHub(IUserProfileService userProfileService, IVideoApiClient videoApiClient, ILogger<ChatHub> logger)
+        {
+            _userProfileService = userProfileService;
+            _videoApiClient = videoApiClient;
+            _logger = logger;
+        }
+        public override async Task OnConnectedAsync()
+        {
+            var userName = await GetObfuscatedUsernameAsync(Context.User.Identity.Name);
+            _logger.LogTrace($"Connected to chat hub server-side: { userName } ");
+            var conferences = await GetConferencesForUser(Context.User.Identity.Name);
+            AddToConferences(conferences);
+
+            await base.OnConnectedAsync();
+        }
+
+        public override async Task OnDisconnectedAsync(Exception exception)
+        {
+            var userName = await GetObfuscatedUsernameAsync(Context.User.Identity.Name);
+            _logger.LogCritical(exception, $"Disconnected from chat hub server-side: { userName } ");
+            var conferences = await GetConferencesForUser(Context.User.Identity.Name);
+            RemoveFromConferences(conferences);
+
+            await base.OnDisconnectedAsync(exception);
+        }
+
+        private async Task<List<ConferenceSummaryResponse>> GetConferencesForUser(string userName)
+        {
+            var conferences = await _videoApiClient.GetConferencesTodayAsync();
+
+            var isAdmin = await IsVhOfficerAsync(Context.User.Identity.Name);
+            if (isAdmin)
+            {
+                return conferences;
+            }
+
+            return conferences.Where(c =>
+                    c.Participants.Any(
+                        p => p.User_role == UserRole.Judge
+                            && p.Username.Equals(userName, StringComparison.InvariantCultureIgnoreCase)))
+                .ToList();
```

Comment:
No need to Tolist it as the consumer of the method just uses it as IEnumerable..

Reply:
ammended